File: app/src/main/java/com/attilapalfi/exceptional/ui/main/MainFragment.java
package com.attilapalfi.exceptional.ui.main;

import javax.inject.Inject;

import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;
import com.attilapalfi.exceptional.R;
import com.attilapalfi.exceptional.dependency_injection.Injector;
import com.attilapalfi.exceptional.interfaces.FirstStartFinishedListener;
import com.attilapalfi.exceptional.interfaces.PointChangeListener;
import com.attilapalfi.exceptional.persistence.FriendStore;
import com.attilapalfi.exceptional.persistence.ImageCache;
import com.attilapalfi.exceptional.persistence.MetadataStore;

/**
 */
public class MainFragment extends Fragment implements FirstStartFinishedListener, PointChangeListener {
    private View view;
    @Inject
    FriendStore friendStore;
    @Inject ImageCache imageCache;
    @Inject MetadataStore metadataStore;

    @Override
    public void onCreate( @Nullable Bundle savedInstanceState ) {
        super.onCreate( savedInstanceState );
        Injector.INSTANCE.getApplicationComponent().inject( this );
    }

    @Nullable
    @Override
    public View onCreateView( LayoutInflater inflater, ViewGroup container,
                              Bundle savedInstanceState ) {
        metadataStore.addFirstStartFinishedListener( this );
        metadataStore.addPointChangeListener( this );
        view = inflater.inflate( R.layout.fragment_main, container, false );
        return view;
    }

    @Override
    public void onResume( ) {
        super.onResume();
        if ( metadataStore.isFirstStartFinished() ) {
            setViews();
        }
    }

    @Override
    public void onDestroyView( ) {
        super.onDestroyView();
        metadataStore.removeFirstStartFinishedListener( this );
        metadataStore.removePointChangeListener( this );
    }

    private void setViews( ) {
        if ( metadataStore.getUser() != null ) {
            setImageView();
            setNameView();
            setPointView();
        }
    }

    private void setImageView( ) {
        ImageView imageView = (ImageView) view.findViewById( R.id.myMainImageView );
        imageCache.setImageToView( metadataStore.getUser(), imageView );
    }

    private void setNameView( ) {
        TextView nameView = (TextView) view.findViewById( R.id.mainNameTextView );
        String nameText = getResources().getString( R.string.main_welcome_before_name )
                + " " + metadataStore.getUser().getFirstName().trim()
                + "!";
        nameView.setText( nameText );
    }

    private void setPointView( ) {
        TextView pointView = (TextView) view.findViewById( R.id.mainPointTextView );
        String pointText = getString( R.string.main_point_view_pre ) + " "
                + metadataStore.getPoints()
                + " " + getString( R.string.main_point_view_post );
        pointView.setText( pointText );
    }

    @Override
    public void onFirstStartFinished( boolean state ) {
        if ( state ) {
            setViews();
        }
    }

    @Override
    public void onPointsChanged( ) {
        setPointView();
    }
}

File: app/src/main/java/com/attilapalfi/exceptional/ui/main/main_page/MainFragment.kt
package com.attilapalfi.exceptional.ui.main.main_page

import android.os.Bundle
import android.support.v4.app.Fragment
import android.support.v7.widget.LinearLayoutManager
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import com.attilapalfi.exceptional.R
import com.attilapalfi.exceptional.dependency_injection.Injector
import com.attilapalfi.exceptional.interfaces.FriendChangeListener
import com.attilapalfi.exceptional.interfaces.GlobalThrowCountChangeListener
import com.attilapalfi.exceptional.interfaces.PointChangeListener
import com.attilapalfi.exceptional.persistence.ExceptionTypeStore
import com.attilapalfi.exceptional.persistence.FriendStore
import com.attilapalfi.exceptional.persistence.ImageCache
import com.attilapalfi.exceptional.persistence.MetadataStore
import com.attilapalfi.exceptional.rest.StatSupplier
import com.attilapalfi.exceptional.ui.main.main_page.recycler_adapter.MainAdapter
import com.attilapalfi.exceptional.ui.main.main_page.recycler_model.FriendPointsChartModel
import com.attilapalfi.exceptional.ui.main.main_page.recycler_model.TypeThrowChartModel
import com.attilapalfi.exceptional.ui.main.main_page.recycler_model.UserRowModel
import javax.inject.Inject

/**
 */
class MainFragment : Fragment(), PointChangeListener, FriendChangeListener, GlobalThrowCountChangeListener {
    private var myView: View? = null
    @Inject
    lateinit var friendStore: FriendStore
    @Inject
    lateinit var imageCache: ImageCache
    @Inject
    lateinit var metadataStore: MetadataStore
    @Inject
    lateinit var typeStore: ExceptionTypeStore
    @Inject
    lateinit var statSupplier: StatSupplier

    private lateinit var recyclerView: RecyclerView
    private lateinit var adapter: MainAdapter

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        Injector.INSTANCE.applicationComponent.inject(this)
    }

    override fun onCreateView(inflater: LayoutInflater?, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        metadataStore.addPointChangeListener(this)
        friendStore.addFriendChangeListener(this)
        statSupplier.addThrowCountListener(this)
        myView = initViewAndModel(inflater, container)
        return myView
    }

    override fun onDestroyView() {
        super.onDestroyView()
        metadataStore.removePointChangeListener(this)
        friendStore.removeFriendChangeListener(this)
        statSupplier.removeThrowCountListener(this)
    }

    private fun initViewAndModel(inflater: LayoutInflater?, container: ViewGroup?): View? {
        if (inflater != null && container != null) {
            val fragmentView = initViewAndRecycler(container, inflater)
            initAdapter()
            return fragmentView
        }
        return null
    }

    private fun initViewAndRecycler(container: ViewGroup?, inflater: LayoutInflater): View? {
        val fragmentView = inflater.inflate(R.layout.fragment_main, container, false);
        recyclerView = fragmentView.findViewById(R.id.main_recycler_view) as RecyclerView;
        recyclerView.layoutManager = LinearLayoutManager(activity)
        return fragmentView
    }

    private fun initAdapter() {
        val adapterData = listOf(UserRowModel(metadataStore.user), FriendPointsChartModel(), TypeThrowChartModel())
        adapter = MainAdapter(adapterData, recyclerView)
        recyclerView.adapter = adapter
    }

    override fun onPointsChanged() {
        adapter.notifyItemChanged(0, 1)
    }

    override fun onFriendsChanged() {
        adapter.notifyItemRangeChanged(0, 1)
    }

    override fun onGlobalThrowCountChanged() {
        adapter.notifyItemChanged(2)
    }
}
